crates/huub/src/constraints/int_value_precede.rs

```diff
@@ -0,0 +1,1144 @@
+			if actions.get_trailed_int(self.first[low]) == i as IntVal {
+				self.propagate_lower_bound(actions, i, low)?;
+			}
+			if i as IntVal <= actions.get_trailed_int(self.last[low])
+				&& actions.check_int_in_domain(v, self.values[low - 1])
+			{
+				let _ = actions.set_trailed_int(self.last[low], i as IntVal);
+				low -= 1;
+			}
+			if low == 0 {
+				break;
+			}
+		}
+
+		self.initialized = true;
+		Ok(())
+	}
+
+	/// Propagate a lower bound by excluding all elements outside values, and
+	/// values with lower index.
+	fn propagate_lower_bound<P: PropagationActions>(
+		&self,
+		actions: &mut P,
+		i: usize,
+		j: usize,
+	) -> Result<(), Conflict> {
+		let lb = actions.get_int_lower_bound(self.vars[i]);
+		let ub = actions.get_int_upper_bound(self.vars[i]);
+		if lb < self.min_val {
+			actions.set_int_lower_bound(self.vars[i], self.min_val, |a: &mut P| {
+				self.explain_lower(a, i, j)
+			})?;
+		}
+		if ub > self.max_val {
+			actions.set_int_upper_bound(self.vars[i], self.max_val, |a: &mut P| {
+				self.explain_lower(a, i, j)
+			})?;
+		}
+		let mut h = max(lb, self.min_hole);
+		while ((h - self.min_hole) as usize) < self.next_hole.len() {
+			h = self.next_hole[(h - self.min_hole) as usize];
+			if h > ub {
+				break;
+			}
+			if actions.check_int_in_domain(self.vars[i], h) {
+				actions.set_int_not_eq(self.vars[i], h, |a: &mut P| self.explain_lower(a, i, j))?;
+			}
+			h += 1;
+		}
+		for k in 0..j - 1 {
+			if actions.check_int_in_domain(self.vars[i], self.values[k]) {
+				actions.set_int_not_eq(self.vars[i], self.values[k], |a: &mut P| {
+					self.explain_lower(a, i, j)
+				})?;
+			}
+		}
+		Ok(())
+	}
+
+	/// Propagate an upper bound by removing all values with higher index.
+	fn propagate_upper_bound<P: PropagationActions>(
+		&self,
+		actions: &mut P,
+		i: usize,
+		j: usize,
+	) -> Result<(), Conflict> {
+		for k in j..self.values.len() {
+			if actions.check_int_in_domain(self.vars[i], self.values[k]) {
+				actions.set_int_not_eq(self.vars[i], self.values[k], |a: &mut P| {
+					self.explain_upper(a, i, k)
+				})?;
+			}
+		}
+		Ok(())
+	}
+
+	/// Iteratively repairs the lower bounds starting with k, only iterates as far
+	/// as necessary.
+	fn repair_lower<P: PropagationActions>(
+		&self,
+		actions: &mut P,
+		mut k: usize,
+	) -> Result<(usize, usize), Conflict> {
+		let mut i = actions.get_trailed_int(self.last[k]);
+		while k > 0 {
+			if actions.check_int_in_domain(self.vars[i as usize], self.values[k - 1]) {
+				let _ = actions.set_trailed_int(self.last[k], i);
+				if actions.get_trailed_int(self.first[k]) == i {
+					self.propagate_lower_bound(actions, i as usize, k)?;
+				}
+				k -= 1;
+				if actions.get_trailed_int(self.last[k]) < i {
+					return Ok((i as usize, k + 1));
+				}
+			}
+
+			i -= 1;
+			if i < 0 {
+				self.propagate_lower_bound(actions, 0, k)?;
+				return Ok((0, k)); //todo There is a conflict now, but it might need propagation to trigger
```

Comment:
How hard would it be to explicitly create an `Conflict` to return? Would we have to copy `propagate_lower_bound`?